src/server/stream_family.cc

```diff
@@ -1328,6 +1332,101 @@ OpResult<uint32_t> OpAck(const OpArgs& op_args, string_view key, string_view gna
   return acknowledged;
 }
 
+OpResult<ClaimInfo> OpAutoClaim(const OpArgs& op_args, string_view key, const ClaimOpts& opts) {
+  OpResult<pair<stream*, streamCG*>> cgr_res = FindGroup(op_args, key, opts.group);
+  if (!cgr_res)
+    return cgr_res.status();
+  auto [stream, group] = *cgr_res;
+
+  if (stream == nullptr || group == nullptr) {
+    return OpStatus::KEY_NOTFOUND;
+  }
+
+  streamConsumer* consumer = nullptr;
+  // from Redis spec on XAutoClaim:
+  // https://redis.io/commands/xautoclaim/
+  // The maximum number of pending entries that the command scans is the product of
+  // multiplying <count>'s value by 10 (hard-coded).
+  int64_t attempts = opts.count * 10;
+
+  unsigned char start_key[sizeof(streamID)];
+  streamID start_id = opts.start;
+  streamEncodeID(start_key, &start_id);
+  raxIterator ri;
+  raxStart(&ri, group->pel);
+  raxSeek(&ri, ">=", start_key, sizeof(start_key));
+
+  ClaimInfo result;
+  result.justid = (opts.flags & kClaimJustID);
+
+  auto now = GetCurrentTimeMs();
+  int count = opts.count;
+  while (attempts-- && count && raxNext(&ri)) {
+    streamNACK* nack = (streamNACK*)ri.data;
+
+    streamID id;
+    streamDecodeID(ri.key, &id);
+
+    if (!streamEntryExists(stream, &id)) {
+      // todo: propagate this change
+      raxRemove(group->pel, ri.key, ri.key_len, nullptr);
+      raxRemove(nack->consumer->pel, ri.key, ri.key_len, nullptr);
+      streamFreeNACK(nack);
+      result.deleted_ids.push_back(id);
+      raxSeek(&ri, ">=", ri.key, ri.key_len);
+      continue;
+    }
+
+    if (opts.min_idle_time) {
+      mstime_t this_idle = now - nack->delivery_time;
+      if (this_idle < opts.min_idle_time)
+        continue;
+    }
+
+    op_args.shard->tmp_str1 =
+        sdscpylen(op_args.shard->tmp_str1, opts.consumer.data(), opts.consumer.size());
+    if (consumer == nullptr) {
+      consumer = streamLookupConsumer(group, op_args.shard->tmp_str1, SLC_DEFAULT);
+      if (consumer == nullptr) {
+        consumer = streamCreateConsumer(group, op_args.shard->tmp_str1, nullptr, 0, SCC_DEFAULT);
+      }
+    }
+
+    if (nack->consumer != consumer) {
+      if (nack->consumer) {
+        raxRemove(nack->consumer->pel, ri.key, ri.key_len, nullptr);
+      }
+    }
+
+    nack->delivery_time = now;
+    if (!result.justid) {
+      nack->delivery_count++;
+    }
+
+    if (nack->consumer != consumer) {
+      raxInsert(consumer->pel, ri.key, ri.key_len, nack, nullptr);
+      nack->consumer = consumer;
+    }
+
+    AppendClaimResultItem(result, stream, id);
+    count--;
+
+    // todo: propagate this change.
```

Comment:
Is this something that will be covered on this PR?

Reply:
this has been covered, the comment will be removed.. thanks for catching this.